hazelcast/src/main/java/com/hazelcast/cache/impl/TenantContextual.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl;
+
+import com.hazelcast.spi.tenantcontrol.TenantControl;
+import com.hazelcast.spi.tenantcontrol.TenantControl.Closeable;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+/**
+ * Represents a value that requires tenant control context to be accessed
+ *
+ * @author lprimak
+ * @param <T> object type
+ */
+public class TenantContextual<T> {
+    private T contextual;
+    private volatile boolean initialized;
+    private final Supplier<T> initFunction;
+    private final Supplier<Boolean> existsFunction;
+    private final TenantControl tenantControl;
+    private final Lock lock = new ReentrantLock();
+
+    public TenantContextual(Supplier<T> initFunction, Supplier<Boolean> existsFunction, TenantControl tenantControl) {
+        this.initFunction = initFunction;
+        this.existsFunction = existsFunction;
+        this.tenantControl = tenantControl;
+    }
+
+    /**
+     *
+     * @return underlying object, initialize within Tenant Control when necessary
+     */
+    public T get() {
+        boolean localInitialized = this.initialized;
+        if (!localInitialized) {
+            try {
+                lock.lock();
```

Comment:
A lock always needs to be acquired outside the try block.

Reply:
good comment. done.